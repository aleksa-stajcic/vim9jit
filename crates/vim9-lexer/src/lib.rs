#![allow(dead_code)]

use std::collections::VecDeque;
use std::fmt::Debug;

#[derive(Clone, PartialEq)]
pub struct Span {
    start_row: usize,
    start_col: usize,
    end_row: usize,
    end_col: usize,
}

impl Debug for Span {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "({},{})->({},{})",
            self.start_row, self.start_col, self.end_row, self.end_col
        )
    }
}

#[derive(Clone, PartialEq)]
pub struct Token {
    pub kind: TokenKind,
    pub text: String,
    pub span: Span,
}

impl Debug for Token {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Token({:?}, {:?}, {:?})", self.kind, self.text, self.span)
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum TokenKind {
    Illegal,
    EndOfFile,

    // TODO: Is this crazy for this lang??
    EndOfLine,
    Comment,

    // Identifiers and literals
    Identifier,
    Integer,

    // Reserved Tokens
    True,
    False,
    Null,

    // Operators
    Equal,
    Plus,
    Minus,
    Mul,
    Div,
    PlusEquals,
    MinusEquals,
    MulEquals,
    DivEquals,

    LessThan,
    LessThanOrEqual,
    GreaterThan,
    GreaterThanOrEqual,
    EqualTo,
    NotEqualTo,
    Or,
    And,

    // Delimiters
    Comma,
    Colon,

    SingleQuoteString,
    DoubleQuoteString,

    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    LeftBracket,
    RightBracket,
}

pub struct Lexer {
    pub input: String,
    pub position: usize,
    pub read_position: usize,
    pub ch: Option<char>,

    chars: Vec<char>,
}

impl Debug for Lexer {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "Lexer {{ position: {}, read_position: {}, ch: {:?} }}",
            self.position, self.read_position, self.ch
        )
    }
}

impl Lexer {
    fn new(input: &str) -> Self {
        let input = input.to_string();
        let chars = input.clone().chars().collect();
        Self {
            input,
            position: 0,
            read_position: 0,
            ch: None,
            chars,
        }
    }

    fn make_span(&self, start: usize, end: usize) -> Span {
        // TODO: This is broken for doing stuff with empty strings... :'(
        assert!(
            start <= end,
            "start must be less than end. {} {}\n{:#?}",
            start,
            end,
            self
        );

        let mut span = Span {
            start_row: 0,
            start_col: 0,
            end_row: 0,
            end_col: 0,
        };

        let mut row = 0;
        let mut row_start = 0;
        for (idx, ch) in self.input.chars().enumerate() {
            if idx == start {
                span.start_row = row;
                span.start_col = idx - row_start;
            }

            if idx == end {
                span.end_row = row;
                span.end_col = idx - row_start;
                return span;
            }

            if ch == '\n' {
                row += 1;
                row_start = idx + 1;
            }
        }

        unreachable!(
            "Should always fine the start and end for a span:\n{} {} {}",
            start,
            end,
            self.input.chars().count()
        )
    }

    pub fn read_char(&mut self) {
        match self.chars.get(self.read_position) {
            Some(&c) => {
                self.ch = Some(c);
                self.position = self.read_position;
                self.read_position += 1;
            }
            None => {
                self.ch = None;
            }
        }
    }

    fn read_number(&mut self) -> Token {
        let position = self.position;
        while let Some(ch) = self.ch && ch.is_numeric(){
            self.read_char();
        }

        Token {
            kind: TokenKind::Integer,
            text: self.chars[position..self.position].iter().collect(),
            span: self.make_span(position, self.position),
        }
    }

    fn read_until(&mut self, until: char, kind: TokenKind) -> Token {
        self.read_char();
        if let Some(ch) = self.ch && ch == until {
            return Token {
                kind,
                text: self.chars[self.position..self.position].iter().collect(),
                span: self.make_span(self.position, self.position ),
            };
        }

        let position = self.position;

        while let Some(ch) = self.ch && ch != until {
            self.read_char();
        }

        Token {
            kind,
            text: self.chars[position..self.position].iter().collect(),
            span: self.make_span(position, self.position - 1),
        }
    }

    fn read_comment(&mut self) -> Token {
        let position = self.position;
        while let Some(ch) = self.ch && ch != '\n' {
            self.read_char();
        }

        Token {
            kind: TokenKind::Comment,
            text: self.chars[position..self.position].iter().collect(),
            span: self.make_span(position, self.position),
        }
    }

    fn read_identifier(&mut self) -> Token {
        let position = self.position;
        while let Some(ch) = self.ch && is_identifier(ch) {
            self.read_char();
        }

        let text: String = self.chars[position..self.position].iter().collect();
        let kind = match text.as_str() {
            "true" => TokenKind::True,
            "false" => TokenKind::False,
            "null" => TokenKind::Null,
            _ => TokenKind::Identifier,
        };

        Token {
            kind,
            text,
            span: self.make_span(position, self.position),
        }
    }

    fn skip_whitespace(&mut self) {
        if self.ch == Some('\n') {
            return;
        }

        while let Some(ch) = self.ch && ch.is_ascii_whitespace() {
            if ch == '\n' {
                return
            }

            self.read_char();
        }
    }

    fn peek_char(&mut self) -> Option<&char> {
        self.chars.get(self.position + 1)
    }

    fn if_peek(&mut self, peeked: char, no: TokenKind, yes: TokenKind) -> Token {
        if let Some(ch) = self.peek_char() {
            if *ch == peeked {
                let position = self.position;
                self.read_char();

                Token {
                    kind: yes,
                    text: self.chars[position..=self.position].iter().collect(),
                    span: self.make_span(position, self.position),
                }
            } else {
                Token {
                    kind: no,
                    text: self.ch.unwrap().to_string(),
                    span: self.make_span(self.position, self.position),
                }
            }
        } else {
            Token {
                kind: TokenKind::EndOfFile,
                text: "".to_string(),
                span: self.make_span(self.position, self.position),
            }
        }
    }

    pub fn next_token(&mut self) -> Token {
        self.skip_whitespace();

        let tok = match self.ch {
            Some(ch) => {
                macro_rules! literal {
                    ($kind:tt) => {
                        Token {
                            kind: TokenKind::$kind,
                            text: ch.to_string(),
                            span: self.make_span(self.position, self.position),
                        }
                    };
                }

                match ch {
                    // Operators that can optionally have an additional equals
                    '=' => self.if_peek('=', TokenKind::Equal, TokenKind::EqualTo),
                    '+' => self.if_peek('=', TokenKind::Plus, TokenKind::PlusEquals),
                    '-' => self.if_peek('=', TokenKind::Minus, TokenKind::MinusEquals),
                    '*' => self.if_peek('=', TokenKind::Mul, TokenKind::MulEquals),
                    '/' => self.if_peek('=', TokenKind::Div, TokenKind::DivEquals),
                    '>' => self.if_peek('=', TokenKind::GreaterThan, TokenKind::GreaterThanOrEqual),
                    '<' => self.if_peek('=', TokenKind::LessThan, TokenKind::LessThanOrEqual),
                    '|' => self.if_peek('|', TokenKind::Illegal, TokenKind::Or),
                    '&' => self.if_peek('&', TokenKind::Illegal, TokenKind::And),

                    ':' => literal!(Colon),
                    '(' => literal!(LeftParen),
                    ')' => literal!(RightParen),
                    '[' => literal!(LeftBracket),
                    ']' => literal!(RightBracket),
                    '{' => literal!(LeftBrace),
                    '}' => literal!(RightBrace),
                    '\n' => literal!(EndOfLine),
                    '#' => self.read_comment(),

                    // TODO: Handle escaped strings.
                    '\'' => self.read_until('\'', TokenKind::SingleQuoteString),
                    '"' => self.read_until('"', TokenKind::DoubleQuoteString),

                    _ => {
                        // Token
                        if is_letter(ch) {
                            return self.read_identifier();
                        } else if ch.is_ascii_digit() {
                            return self.read_number();
                        } else {
                            Token {
                                kind: TokenKind::Illegal,
                                text: ch.to_string(),
                                span: self.make_span(self.position, self.position),
                            }
                        }
                    }
                }
            }
            None => Token {
                kind: TokenKind::EndOfFile,
                text: "".to_string(),
                span: self.make_span(self.position, self.position),
            },
        };

        self.read_char();
        tok
    }
}

fn is_identifier(ch: char) -> bool {
    ch.is_alphanumeric() || ch == '_'
}

fn is_letter(ch: char) -> bool {
    'a' <= ch && ch <= 'z' || 'A' <= ch && ch <= 'Z'
}

fn is_digit(ch: char) -> bool {
    '0' <= ch && ch <= '9'
}

pub fn new_lexer(input: &str) -> Lexer {
    let mut l = Lexer::new(input);
    l.read_char();
    l
}

pub fn snapshot_lexing(input: &str) -> String {
    let mut lexer = new_lexer(input);

    let mut tokens = VecDeque::new();
    loop {
        let tok = lexer.next_token();
        if tok.kind == TokenKind::EndOfFile {
            break;
        }

        tokens.push_back(tok);
    }

    let mut output = String::new();
    for (row, line) in input.lines().enumerate() {
        output += line;
        output += "\n";

        while let Some(tok) = tokens.pop_front() {
            if tok.span.start_row != tok.span.end_row {
                panic!("We haven't handled this yet");
            }

            if tok.span.start_row != row {
                tokens.push_front(tok);
                break;
            }

            output += &" ".repeat(tok.span.start_col);
            output += &"^".repeat(tok.span.end_col + 1 - tok.span.start_col);
            output += &format!(" {:?}", tok);
            output += "\n"
        }
    }

    output
}

#[cfg(test)]
mod test {
    use super::*;

    macro_rules! snapshot {
        ($name:tt, $path:tt) => {
            #[test]
            fn $name() {
                let contents = include_str!($path);
                let mut settings = insta::Settings::clone_current();
                settings.set_snapshot_path("../testdata/output/");
                settings.bind(|| {
                    insta::assert_snapshot!(snapshot_lexing(contents));
                });
            }
        };
    }

    snapshot!(test_lexer_1, "../testdata/snapshots/lexer_1.vim");
    snapshot!(test_comparisons, "../testdata/snapshots/comparisons.vim");
    snapshot!(test_string, "../testdata/snapshots/string.vim");

    // TODO: Check more thoroughly
    snapshot!(test_matchparen, "../testdata/snapshots/matchparen.vim");
}
