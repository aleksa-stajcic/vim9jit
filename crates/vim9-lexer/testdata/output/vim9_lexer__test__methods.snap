---
source: crates/vim9-lexer/src/lib.rs
assertion_line: 936
expression: snapshot_lexing(contents)
---
vim9script
^^^^^^^^^^ Token(Identifier, "vim9script", (0,0)->(0,10))
           Token(EndOfLine, "\n", (0,10)->(0,10))

 Token(EndOfLine, "\n", (1,0)->(1,0))
var l = "hello"->len()
^^^ Token(Identifier, "var", (2,0)->(2,3))
    ^ Token(Identifier, "l", (2,4)->(2,5))
      ^ Token(Equal, "=", (2,6)->(2,7))
         ^^^^ Token(DoubleQuoteString, "hello", (2,9)->(2,13))
               ^^ Token(MethodArrow, "->", (2,15)->(2,17))
                 ^^^ Token(Identifier, "len", (2,17)->(2,20))
                    ^ Token(LeftParen, "(", (2,20)->(2,21))
                     ^ Token(RightParen, ")", (2,21)->(2,22))
                       Token(EndOfLine, "\n", (2,22)->(2,22))

 Token(EndOfLine, "\n", (3,0)->(3,0))
var x = [5, 4, 2, 1, 7, 12, 8]
^^^ Token(Identifier, "var", (4,0)->(4,3))
    ^ Token(Identifier, "x", (4,4)->(4,5))
      ^ Token(Equal, "=", (4,6)->(4,7))
        ^ Token(LeftBracket, "[", (4,8)->(4,9))
         ^ Token(Integer, "5", (4,9)->(4,10))
          ^ Token(Comma, ",", (4,10)->(4,11))
            ^ Token(Integer, "4", (4,12)->(4,13))
             ^ Token(Comma, ",", (4,13)->(4,14))
               ^ Token(Integer, "2", (4,15)->(4,16))
                ^ Token(Comma, ",", (4,16)->(4,17))
                  ^ Token(Integer, "1", (4,18)->(4,19))
                   ^ Token(Comma, ",", (4,19)->(4,20))
                     ^ Token(Integer, "7", (4,21)->(4,22))
                      ^ Token(Comma, ",", (4,22)->(4,23))
                        ^^ Token(Integer, "12", (4,24)->(4,26))
                          ^ Token(Comma, ",", (4,26)->(4,27))
                            ^ Token(Integer, "8", (4,28)->(4,29))
                             ^ Token(RightBracket, "]", (4,29)->(4,30))
                               Token(EndOfLine, "\n", (4,30)->(4,30))
      ->filter((_, x) => x % 2 == 0)
      ^^ Token(MethodArrow, "->", (5,6)->(5,8))
        ^^^^^^ Token(Identifier, "filter", (5,8)->(5,14))
              ^ Token(LeftParen, "(", (5,14)->(5,15))
               ^ Token(LeftParen, "(", (5,15)->(5,16))
                ^ Token(Identifier, "_", (5,16)->(5,17))
                 ^ Token(Comma, ",", (5,17)->(5,18))
                   ^ Token(Identifier, "x", (5,19)->(5,20))
                    ^ Token(RightParen, ")", (5,20)->(5,21))
                      ^^ Token(Arrow, "=>", (5,22)->(5,24))
                         ^ Token(Identifier, "x", (5,25)->(5,26))
                           ^ Token(Percent, "%", (5,27)->(5,28))
                             ^ Token(Integer, "2", (5,29)->(5,30))
                               ^^ Token(EqualTo, "==", (5,31)->(5,33))
                                  ^ Token(Integer, "0", (5,34)->(5,35))
                                   ^ Token(RightParen, ")", (5,35)->(5,36))
                                     Token(EndOfLine, "\n", (5,36)->(5,36))
      ->map((_, y) => y + 1)
      ^^ Token(MethodArrow, "->", (6,6)->(6,8))
        ^^^ Token(Identifier, "map", (6,8)->(6,11))
           ^ Token(LeftParen, "(", (6,11)->(6,12))
            ^ Token(LeftParen, "(", (6,12)->(6,13))
             ^ Token(Identifier, "_", (6,13)->(6,14))
              ^ Token(Comma, ",", (6,14)->(6,15))
                ^ Token(Identifier, "y", (6,16)->(6,17))
                 ^ Token(RightParen, ")", (6,17)->(6,18))
                   ^^ Token(Arrow, "=>", (6,19)->(6,21))
                      ^ Token(Identifier, "y", (6,22)->(6,23))
                        ^ Token(Plus, "+", (6,24)->(6,25))
                          ^ Token(Integer, "1", (6,26)->(6,27))
                           ^ Token(RightParen, ")", (6,27)->(6,28))
                             Token(EndOfLine, "\n", (6,28)->(6,28))
      ->sort()
      ^^ Token(MethodArrow, "->", (7,6)->(7,8))
        ^^^^ Token(Identifier, "sort", (7,8)->(7,12))
            ^ Token(LeftParen, "(", (7,12)->(7,13))
             ^ Token(RightParen, ")", (7,13)->(7,14))
               Token(EndOfLine, "\n", (7,14)->(7,14))

 Token(EndOfLine, "\n", (8,0)->(8,0))
var expr_prec = -1.234->string()
^^^ Token(Identifier, "var", (9,0)->(9,3))
    ^^^^^^^^^ Token(Identifier, "expr_prec", (9,4)->(9,13))
              ^ Token(Equal, "=", (9,14)->(9,15))
                ^ Token(Minus, "-", (9,16)->(9,17))
                 ^^^^^ Token(Float, "1.234", (9,17)->(9,22))
                      ^^ Token(MethodArrow, "->", (9,22)->(9,24))
                        ^^^^^^ Token(Identifier, "string", (9,24)->(9,30))
                              ^ Token(LeftParen, "(", (9,30)->(9,31))
                               ^ Token(RightParen, ")", (9,31)->(9,32))
                                 Token(EndOfLine, "\n", (9,32)->(9,32))

 Token(EndOfLine, "\n", (10,0)->(10,0))
var foo = base->name(args)
^^^ Token(Identifier, "var", (11,0)->(11,3))
    ^^^ Token(Identifier, "foo", (11,4)->(11,7))
        ^ Token(Equal, "=", (11,8)->(11,9))
          ^^^^ Token(Identifier, "base", (11,10)->(11,14))
              ^^ Token(MethodArrow, "->", (11,14)->(11,16))
                ^^^^ Token(Identifier, "name", (11,16)->(11,20))
                    ^ Token(LeftParen, "(", (11,20)->(11,21))
                     ^^^^ Token(Identifier, "args", (11,21)->(11,25))
                         ^ Token(RightParen, ")", (11,25)->(11,26))
                           Token(EndOfLine, "\n", (11,26)->(11,26))
var foo = base->some.name(args)
^^^ Token(Identifier, "var", (12,0)->(12,3))
    ^^^ Token(Identifier, "foo", (12,4)->(12,7))
        ^ Token(Equal, "=", (12,8)->(12,9))
          ^^^^ Token(Identifier, "base", (12,10)->(12,14))
              ^^ Token(MethodArrow, "->", (12,14)->(12,16))
                ^^^^ Token(Identifier, "some", (12,16)->(12,20))
                    ^ Token(Dot, ".", (12,20)->(12,21))
                     ^^^^ Token(Identifier, "name", (12,21)->(12,25))
                         ^ Token(LeftParen, "(", (12,25)->(12,26))
                          ^^^^ Token(Identifier, "args", (12,26)->(12,30))
                              ^ Token(RightParen, ")", (12,30)->(12,31))
                                Token(EndOfLine, "\n", (12,31)->(12,31))
var foo = base->alist[idx](args)
^^^ Token(Identifier, "var", (13,0)->(13,3))
    ^^^ Token(Identifier, "foo", (13,4)->(13,7))
        ^ Token(Equal, "=", (13,8)->(13,9))
          ^^^^ Token(Identifier, "base", (13,10)->(13,14))
              ^^ Token(MethodArrow, "->", (13,14)->(13,16))
                ^^^^^ Token(Identifier, "alist", (13,16)->(13,21))
                     ^ Token(LeftBracket, "[", (13,21)->(13,22))
                      ^^^ Token(Identifier, "idx", (13,22)->(13,25))
                         ^ Token(RightBracket, "]", (13,25)->(13,26))
                          ^ Token(LeftParen, "(", (13,26)->(13,27))
                           ^^^^ Token(Identifier, "args", (13,27)->(13,31))
                               ^ Token(RightParen, ")", (13,31)->(13,32))
                                 Token(EndOfLine, "\n", (13,32)->(13,32))
var foo = base->(getFuncRef())(args)
^^^ Token(Identifier, "var", (14,0)->(14,3))
    ^^^ Token(Identifier, "foo", (14,4)->(14,7))
        ^ Token(Equal, "=", (14,8)->(14,9))
          ^^^^ Token(Identifier, "base", (14,10)->(14,14))
              ^^ Token(MethodArrow, "->", (14,14)->(14,16))
                ^ Token(LeftParen, "(", (14,16)->(14,17))
                 ^^^^^^^^^^ Token(Identifier, "getFuncRef", (14,17)->(14,27))
                           ^ Token(LeftParen, "(", (14,27)->(14,28))
                            ^ Token(RightParen, ")", (14,28)->(14,29))
                             ^ Token(RightParen, ")", (14,29)->(14,30))
                              ^ Token(LeftParen, "(", (14,30)->(14,31))
                               ^^^^ Token(Identifier, "args", (14,31)->(14,35))
                                   ^ Token(RightParen, ")", (14,35)->(14,36))
                                     Token(EndOfLine, "\n", (14,36)->(14,36))

 Token(EndOfLine, "\n", (15,0)->(15,0))
var foo = [slnum, scol]->cursor()
^^^ Token(Identifier, "var", (16,0)->(16,3))
    ^^^ Token(Identifier, "foo", (16,4)->(16,7))
        ^ Token(Equal, "=", (16,8)->(16,9))
          ^ Token(LeftBracket, "[", (16,10)->(16,11))
           ^^^^^ Token(Identifier, "slnum", (16,11)->(16,16))
                ^ Token(Comma, ",", (16,16)->(16,17))
                  ^^^^ Token(Identifier, "scol", (16,18)->(16,22))
                      ^ Token(RightBracket, "]", (16,22)->(16,23))
                       ^^ Token(MethodArrow, "->", (16,23)->(16,25))
                         ^^^^^^ Token(Identifier, "cursor", (16,25)->(16,31))
                               ^ Token(LeftParen, "(", (16,31)->(16,32))
                                ^ Token(RightParen, ")", (16,32)->(16,33))
                                  Token(EndOfLine, "\n", (16,33)->(16,33))

 Token(EndOfLine, "\n", (17,0)->(17,0))
# Testing
^^^^^^^^^ Token(Comment, "# Testing", (18,0)->(18,9))
[slnum, scol]->cursor()
^ Token(LeftBracket, "[", (19,0)->(19,1))
 ^^^^^ Token(Identifier, "slnum", (19,1)->(19,6))
      ^ Token(Comma, ",", (19,6)->(19,7))
        ^^^^ Token(Identifier, "scol", (19,8)->(19,12))
            ^ Token(RightBracket, "]", (19,12)->(19,13))
             ^^ Token(MethodArrow, "->", (19,13)->(19,15))
               ^^^^^^ Token(Identifier, "cursor", (19,15)->(19,21))
                     ^ Token(LeftParen, "(", (19,21)->(19,22))
                      ^ Token(RightParen, ")", (19,22)->(19,23))
                        Token(EndOfLine, "\n", (19,23)->(19,23))

